### PR TITLE
(WIP) feat/heuristic-clique-oracle

### PR DESCRIPTION
**Pull request by SeanAvery**
## Heuristic Clique Oracle ##

Right now, the functionality is appended to the CliqueOracle class. For testing, the oracle is being called from the blockchain view. 

To run: `python casper.py rand --validators 10 --rounds 100` Check console output to see it in action.

This oracle does a greedy search in attempt to find a large clique quickly. It iterates through the graph favoring vertices with the largest degree until it can reduce a clique.

The current implementation does not take into account weight, which I plan on adding in.

The greedy algorithm can also be extended beyond a simple lookup to do a full search on the graph space.

For more details see #134

### CODE REVIEW

File: casper/safety_oracles/clique_oracle.py [hunk 1]
```diff
@@ -112,3 +112,72 @@ def check_estimate_safety(self):
...
+            print("### found a clique of size:", heuristic_min_clique, len(heuristic_min_clique))
+            return heuristic_min_clique
+
+        # recursively call the greedy search
+        return self.greedy_clique_search(graph, heuristic_min_clique, temp_max_degree_node, max_degree_neighbors)
+
+    def find_max_degree_node(self, graph, nodes):
+        """Returns greatest degree node out of list"""
+        max_degree_node = None
+        for node in nodes:
```
Comment (naterush) [inline casper/safety_oracles/clique_oracle.py]: You can use python built-ins here. They are more readable and also faster! Something like: 

`max_degree_node = max(nodes, key=lambda node: graph.degree(node))`

Not sure the above will work b/c I can't get this code running for some reason, but something similar should work :)

File: casper/safety_oracles/clique_oracle.py [hunk 2]
```diff
@@ -112,3 +112,72 @@ def check_estimate_safety(self):
...
+
+        # if graph is empty, break out of function
+        if not nodes:
+            print('### graph is empty, not running clique search')
+            return
+
+        # base case: find largest degree node in graph
+        max_degree_node = self.find_max_degree_node(graph, nodes)
+        if not max_degree_node:
+            raise Exception("### Did not find a node in the graph, no common estimates")
```
Comment (naterush) [inline casper/safety_oracles/clique_oracle.py]: Does this ever trigger? It seems like if we have any nodes (as nodes are "inferred" from edges), we should have some node with a max degree.

File: casper/safety_oracles/clique_oracle.py [hunk 3]
```diff
@@ -112,3 +112,72 @@ def check_estimate_safety(self):
...
+        latest_max_degree_neighbors = graph.neighbors(max_degree_node)
+
+        # if vertex does not have neighbors finish search, the graph is empty
+        if not latest_max_degree_neighbors:
+            print("### vertex has no neighbors")
+            print("### found a clique of size:", heuristic_min_clique, len(heuristic_min_clique))
+            return heuristic_min_clique
+
+        # remove neighbors that are not adjacent to the new max degree vertex
+        for node in max_degree_neighbors:
```
Comment (naterush) [inline casper/safety_oracles/clique_oracle.py]: You can change this to a Python built-in. Something like `difference_update`

Read more here: https://docs.python.org/2/library/sets.html

File: casper/safety_oracles/clique_oracle.py [hunk 4]
```diff
@@ -112,3 +112,72 @@ def check_estimate_safety(self):
...
+        """Returns a minimum bound clique using a greedy search lookup."""
+
+        # collect edges and create graph
+        edges = self._collect_edges()
+        graph = nx.Graph()
+        graph.add_edges_from(edges)
+        nodes = graph.nodes()
+
+        # if graph is empty, break out of function
+        if not nodes:
```
Comment (naterush) [inline casper/safety_oracles/clique_oracle.py]: maybe, if `not any(nodes):`

**Comment by naterush** [comment]
@SeanAvery Much like the Turan's oracle, I think this might be mildly "attackable." If I'm an attacker that controls some portion of the network w/ low weight validators, they can construct a high-degree connected node (that would be first chosen in the greedy search), that ends up not contributing to safety very much b/c of it's low weight.

This disappears once you add in weight considerations!

**Comment by SeanAvery** [comment]
@naterush Completely agree on vulnerability to attacks. i.e. A big hub and spoke resulting in a clique size of 2! Taking into account weight will make it much more expensive to perform such an attack. Still brainstorming on how to properly perform the greedy search w/ weight but i have some ideas. Going to work on rebasing this oracle-- cleaning up code and weight integration.

File: casper/safety_oracles/clique_oracle.py (hunk 3 above)
Comment (SeanAvery) [reply casper/safety_oracles/clique_oracle.py]: Oh cool. Yeah for sure, will refactor to use built-ins. Starting to get this whole python thing.. lol.

**Pull request closed without merging**